### DISCUSSION TITLE
What is the difference between L1 and L2 loss functions?

### DISCUSSION THREAD

**Discussion by aasimansari1** [aasimansari1/ml-interview-prep#21] [Q&A] [answered]
What is the difference between L1 and L2 loss functions?

**Comment by aasimansari1** [comment, marked as answer]
**Loss functions** measure how wrong predictions are — L1 and L2 behave very differently with outliers.

## L2 Loss (MSE — Mean Squared Error)
```python
import torch.nn as nn
import numpy as np

# L2: squares the error — outliers get huge penalty
mse = nn.MSELoss()
loss = mse(predictions, targets)
# = mean((pred - true)^2)
```
✅ **Differentiable everywhere** — smooth gradient descent
✅ **Unique minimum** — stable optimization
❌ **Sensitive to outliers** — one large error dominates everything

## L1 Loss (MAE — Mean Absolute Error)
```python
mae = nn.L1Loss()
loss = mae(predictions, targets)
# = mean(|pred - true|)
```
✅ **Robust to outliers** — large errors don't dominate
✅ **Better when data is noisy**
❌ **Non-differentiable at 0** — gradient is undefined at perfect prediction

## Huber Loss (best of both)
```python
huber = nn.HuberLoss(delta=1.0)  # L2 for small errors, L1 for large
```
✅ Smooth like L2, robust like L1 — use this when you have some outliers

## Quick decision:
| Scenario | Use |
|----------|-----|
| Clean data, normal errors | **MSE (L2)** |
| Noisy data, outliers present | **MAE (L1)** |
| Mix of clean + some outliers | **Huber** |
| Classification | **CrossEntropy** (not L1/L2) |

**Interview insight**: MSE penalizes large errors quadratically — predicting 10 when truth is 0 gets 100x more penalty than predicting 1. MAE treats all errors linearly.
